File: Linked List/82.remove-duplicates-from-sorted-list-ii.py
#
# @lc app=leetcode id=82 lang=python3
#
# [82] Remove Duplicates from Sorted List II
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def deleteDuplicates(self, head: ListNode) -> ListNode:
        p = ListNode(0)
        p.next, head = head, p
        while p.next:
            left = p.next
            right = left
            while right.next and right.next.val == left.val:
                right = right.next
            if left == right:
                p = p.next
            else:
                p.next = right.next
        return head.next
# @lc code=end

File: Linked List/82.remove-duplicates-from-sorted-list-ii.cpp
/*
 * @lc app=leetcode id=82 lang=cpp
 *
 * [82] Remove Duplicates from Sorted List II
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* deleteDuplicates(ListNode* head) {
        auto dummy = new ListNode(-100);
        dummy->next = head;
        auto fast = dummy, slow = dummy;
        while(fast && fast->next)
        {
            if(fast->val != fast->next->val){
                fast = fast->next;
                if(fast && fast->next && fast->val != fast->next->val)
                    slow = fast;
            }
            else{
                while(fast && fast->next && fast->val == fast->next->val)
                    fast = fast->next;
                fast = fast->next;
                slow->next = fast;
                if(fast && fast->next && fast->next->val != fast->val) slow = fast;
            }
            
        }   
        return dummy->next;
    }
};
// @lc code=end

